File: src/main/python/medium/Q0017.py
"""
17. Letter Combinations of a Phone Number
https://leetcode.com/problems/letter-combinations-of-a-phone-number/
"""


# Time: O(n 4^n), Space: O(n log n), where n is the length of `digits`.
def letter_combinations(digits: str) -> list[str]:
    digit_to_chars = {
        '2': ['a', 'b', 'c'],
        '3': ['d', 'e', 'f'],
        '4': ['g', 'h', 'i'],
        '5': ['j', 'k', 'l'],
        '6': ['m', 'n', 'o'],
        '7': ['p', 'q', 'r', 's'],
        '8': ['t', 'u', 'v'],
        '9': ['w', 'x', 'y', 'z'],
    }
    result: list[str] = []

    def backtrack(i: int, s: string):
        if len(s) == len(digits):
            result.append(s)
            return
        d = digits[i]
        for c in digit_to_chars[d]:
            backtrack(i + 1, s + c)

    if digits != "":
        backtrack(0, "")
    return result

File: src/main/java/medium/Q0017.java
package medium;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 17. Letter Combinations of a Phone Number
 * https://leetcode.com/problems/letter-combinations-of-a-phone-number/
 */
public class Q0017 {

    private Map<Character, String> digitLetters = new HashMap<>(Map.of(
            '2', "abc",
            '3', "def",
            '4', "ghi",
            '5', "jkl",
            '6', "mno",
            '7', "pqrs",
            '8', "tuv",
            '9', "wxyz"
    ));
    private List<String> result = new ArrayList<>();
    private String digits;

    private void backtrack(int i, String s) {
        if (s.length() == this.digits.length()) {
            this.result.add(s);
            return;
        }
        var digit = this.digits.charAt(i);
        for (var c : this.digitLetters.get(digit).toCharArray()) {
            backtrack(i + 1, s + c);
        }
    }

    // Time: O(n 4^n), Space: O(n log n), where n is the length of `digits`.
    private List<String> letterCombinations(String digits) {
        this.digits = digits;
        if (!digits.isEmpty()) {
            backtrack(0, new String());
        }
        return this.result;
    }
}
